### PR TITLE
ENH: allow dataframe get to take an axis argument

### LINKED ISSUE #6703: ENH: allow get to take an axis argument
http://stackoverflow.com/questions/22636487/selection-in-multi-indexed-dataframe-when-key-does-not-exists/22637762#22637762

```
In [23]: df = DataFrame({'a':[1,2,3],'b':[1,2,3],'c':[1,2,3]})

In [24]: x = df.set_index(['a','b'])

In [25]: x
Out[25]: 
     c
a b   
1 1  1
2 2  2
3 3  3

[3 rows x 1 columns]

In [26]: x.T.get((1,1))
Out[26]: 
c    1
Name: (1, 1), dtype: int64

In [27]: x.T.get((1,5),default='foo')
Out[27]: 'foo'
```

``x.get((1,5),default='foo',axis=0)`` should work (still need the ``info_axis`` to be the default though)

### PR DESCRIPTION
**Pull request by nbonnotte**
closes #6703

I would have like to have a solution also working for panels, but I've never used them, so I'm not sure how it should behave.

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -577,6 +577,30 @@ def _repr_html_(self):
         else:
             return None
 
+    def get(self, key, default=None, axis='columns'):
```
Comment (jreback) [inline pandas/core/frame.py]: just modify the one in ``core/generic.py``

File: pandas/core/frame.py [hunk 2]
```diff
@@ -577,6 +577,30 @@ def _repr_html_(self):
...
+            Axis in which the key is to be found. Default is 'columns'
+        key : object
+
+        Returns
+        -------
+        value : type of items contained in object
+        """
+        df = self
+        axis = self._get_axis_number(axis)
+        if axis == 0:
```
Comment (jreback) [inline pandas/core/frame.py]: to get an arbitrary just do this:

```
slices = [ slice(None) ] * self.ndim
slices[axis] = key
try:
    return self.loc[tuple(slices)]
except (KeyError, ......):
    return default
```

File: pandas/tests/test_frame.py [hunk 3]
```diff
@@ -149,6 +149,12 @@ def test_get(self):
             result = df.get(None)
             self.assertIsNone(result)
 
+        # GH 6703
```
Comment (jreback) [inline pandas/tests/test_frame.py]: this should be in ``tests/test_generic.py``. In fact ideally we would move the tests from where they are now (test_series/frame) and move them here (in the appopriate test classes).

**Comment by nbonnotte** [comment]
Ok, I used your suggestion and put the tests in `tests/test_generic.py`... but it still is the same test. I should probably make more tests, in particular to test what happens for Panel.

(Sorry, Atom couldn't help but by itself remove some spaces that were at the end of their lines)

**Comment by nbonnotte** [comment]
I don't understand, why is the build failing?

**Comment by maximilianr** [comment]
If you go to 'Details', you can get the build's output. In this case:
```
======================================================================
FAIL: test_getitem_get (pandas.tests.test_series.TestSeries)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/pydata/pandas/pandas/tests/test_series.py", line 1339, in test_getitem_get
    self.series.get(-1), self.series.get(self.series.index[-1]))
AssertionError: None != -0.2267325702945841
```

**Comment by nbonnotte** [comment]
Uh, yes, thanks.

I think I've identified what is going on, but I'm not sure on how to proceed:

```
In [56]: pd.Series([0,1], index=[0,1]).get(-1)

In [57]: pd.Series([0,1], index=['a','b']).get(-1)
Out[57]: 1
```
So, basically, the result of the `get` method for a series depends on the type of the index... Which is not the case for a dataframe:
```
In [63]: pd.DataFrame([[0,1],[2,3]], columns=['a','b']).get(-1) is None
Out[63]: True
```

I could redefine `Series.get` to supersede `NDFrame.get`... but I don't know if that's a good solution. Maybe there are other classes inheriting from `NDFrame` that work like `Series`?

Here is what I'm proposing: if the solution based on the slice and `loc` fail, and `axis` is `None`, we revert to the old implementation. Thus we maintain a backward compatibility...

File: pandas/core/generic.py [hunk 4]
```diff
@@ -1062,23 +1062,51 @@ def _indexer(self):
...
+            The axis to filter on. By default this is the info axis. The "info
+            axis" is the axis that is used when indexing with ``[]``. For
+            example, ``df = DataFrame({'a': [1, 2, 3, 4]]})
 
         Returns
         -------
-        value : type of items contained in object
+        value : type of items contained in object, or default
         """
+        # special case (GH 5652)
```
Comment (jreback) [inline pandas/core/generic.py]: not needed, this just generates a ``ValueError`` which by definition returns ``default``

File: pandas/core/generic.py [hunk 5]
```diff
@@ -1062,23 +1062,51 @@ def _indexer(self):
...
+        if key is None:
+            return default
+        # general case
+        if axis is None:
+            axis = self._info_axis_number
+        else:
```
Comment (jreback) [inline pandas/core/generic.py]: more idiomatic is

```
if axis is None:
    axis = self._info_axis_number
axis = self._get_axis_number(axis)
```

File: pandas/core/generic.py [hunk 6]
```diff
@@ -1062,23 +1062,51 @@ def _indexer(self):
...
+        slices = [slice(None)] * self.ndim
+        slices[axis] = key
         try:
-            return self[key]
+            return self.loc[tuple(slices)]
         except (KeyError, ValueError, IndexError):
-            return default
+            pass
+        # Two possibilities:
+        # 1) the key is not present, and we should return default
```
Comment (jreback) [inline pandas/core/generic.py]: way to convoluted. show an example of this case.

File: pandas/tests/test_generic.py [hunk 7]
```diff
@@ -99,6 +99,14 @@ def test_rename(self):
 
         # multiple axes at once
 
+    def test_get(self):
+        # GH 6703
+        # testing the axis parameter
```
Comment (jreback) [inline pandas/tests/test_generic.py]: need generic testing on Series,DataFrame,Panel for all axes.

File: pandas/core/generic.py (hunk 4 above)
Comment (nbonnotte) [inline pandas/core/generic.py]: That generates a `TypeError` ("cannot use label indexing with a null key"). But you're right, I can also catch it below...

**Comment by shoyer** [comment]
I'm not sure this is a good idea. Get is a standard method for mapping types and we already have a lot of indexing methods.

**Comment by jreback** [comment]
yeh, this is getting into too many edge cases. Lets just keep ``.get`` simple.

**Pull request closed without merging**